iati/core/tests/test_data.py

```diff
@@ -406,3 +407,95 @@ def test_dataset_xml_str_source_around_line_invalid_context_lines(self, invalid_
         for line_num in range(0, num_lines_xml):
             with pytest.raises(TypeError):
                 data.source_around_line(line_num, invalid_value)
+
+
+class TestDatasetVersionDetection(object):
+    """A container for tests relating to detecting the version of a Dataset."""
+
+    @pytest.mark.parametrize("version", iati.core.default.get_versions_by_integer()[1])
+    def test_detect_version_v1_simple(self, version):
+        """Check that a version 1 dataset is detected correctly."""
+        data = iati.core.Dataset("""
+        <iati-activities version="{0}">
+            <iati-activity version="{0}"></iati-activity>
+            <iati-activity version="{0}"></iati-activity>
+            <iati-activity version="{0}"></iati-activity>
+        </iati-activities>
+        """.format(version))
+        result = data.version
+
+        assert result == version
+
+    def test_detect_version_explicit_iati_activities_mismatch_explicit_iati_activity(self):
+        data = iati.core.Dataset("""
+        <iati-activities version="1.02">
+            <iati-activity version="1.02"></iati-activity>
+            <iati-activity version="1.02"></iati-activity>
+            <iati-activity version="1.03"></iati-activity>
+        </iati-activities>
+        """)
+        result = data.version
+
+        assert result is None
+
+    def test_detect_version_implicit_iati_activities_matches_implicit_iati_activity(self):
+        data = iati.core.Dataset("""
+        <iati-activities>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+        </iati-activities>
+        """)
+        result = data.version
+
+        assert result == '1.01'
+
+    def test_detect_version_explicit_iati_activities_matches_implicit_iati_activity(self):
+        data = iati.core.Dataset("""
+        <iati-activities version='1.01'>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+        </iati-activities>
+        """)
+        result = data.version
+
+        assert result == '1.01'
+
+    def test_detect_version_explicit_iati_activities_mismatch_implicit_iati_activity(self):
+        data = iati.core.Dataset("""
+        <iati-activities version='1.02'>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+            <iati-activity></iati-activity>
+        </iati-activities>
+        """)
+        result = data.version
+
+        assert result is None
+
+    def test_detect_version_imlicit_iati_activities_mismatch_explicit_iati_activity(self):
+        data = iati.core.Dataset("""
+        <iati-activities>
+            <iati-activity version="1.02"></iati-activity>
```

Comment:
What happens if there's just `1.01` on a singular?

Reply:
Do you mean
```
<iati-activities>
   <iati-activity version="1.02"></iati-activity>
</iati-activities>
```

??

Reply:
That probably plus a couple of blank singulars for consistency. Basically, what this test is checking but with `1.01` rather than `1.02`.